File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(
            (a,b) ->
            {
                if(a[0] != b[0])
                {
                    return b[0] - a[0]; // descending
                }
                return a[1] - b[1]; // ascending
            }
        );
        ArrayList<Integer> ans = new ArrayList<>();

        for(int i=0;i<k;i++)
        {
            pq.add(new int[]{nums[i],i});
        }

        ans.add(pq.peek()[0]);
        int minIndex = 0;

        for(int i=k;i<nums.length;i++)
        {
            pq.add(new int[] {nums[i],i});
            while(pq.peek()[1] <= minIndex)
            {
                pq.poll();
            }
            minIndex++;
            ans.add(pq.peek()[0]);
        }

        int [] res =  new int[ans.size()];
        for(int i=0;i<ans.size();i++)
        {
            res[i] = ans.get(i);
        }
        
        return res;

    }
}

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:

        pq = []
        ans= []

        for i in range (0,k,1):
            heapq.heappush(pq,(-nums[i],i))

        ans.append(-pq[0][0])
        minIndex = 0

        for i in range (k,len(nums),1):
            heapq.heappush(pq,(-nums[i],i))
            while(pq[0][1] <= minIndex):
                heapq.heappop(pq)
            minIndex += 1
            ans.append(-pq[0][0])
        
        return ans